### PR TITLE
SUP-16283: Make bulk request loader follow the limits in the edge cases

### PR DESCRIPTION
**Pull request by plyhun**
## Abstract

There exists a possibility of DDoSing the Elasticsearch server, if a network glitch makes Mesh collecting unlimited ES requests, which are then pushed for the processing all at once.

### General

* [x] Added abstract that describes the change
* [x] Added changelog entry to `/CHANGELOG.adoc`
* [x] Ensured that the change is covered by tests
* [x] Ensured that the change is documented in the docs

### On API Changes

* [x] Checked if the changes are breaking or not
* [x] Added GraphQL API if applicable
* [x] Added Elasticsearch mapping if applicable

### CODE REVIEW

File: elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java [hunk 1]
```diff
@@ -85,12 +89,22 @@ public void drain() {
...
 						log.trace("Emitting bulk of size {} to subscriber", bulkableRequests.size());
-						BulkRequest request = new BulkRequest(bulkableRequests.asList());
-						bulkableRequests.clear();
-						if (log.isDebugEnabled()) {
-							log.debug("Sending bulk to elasticsearch:\n{}", request);
-						}
-						subscriber.onNext(request);
+						do {
+							Long[] bulkLength = new Long[] { 0L };
+							List<Bulkable> requests = IntStream.range(0, requestLimit)
```
Comment (plyhun) [inline elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java:94]: Obey number limit.

File: elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java [hunk 2]
```diff
@@ -85,12 +89,22 @@ public void drain() {
...
+								.mapToObj(index -> bulkableRequests.poll())
+								.takeWhile(maybeBulkable -> maybeBulkable
+										.filter(bulkable -> bulkLength[0] < lengthLimit)
```
Comment (plyhun) [inline elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java:97]: Obey size limit.

File: tests/tests-core/src/main/java/com/gentics/mesh/search/ElasticSearchProviderTimeoutTest.java [hunk 3]
```diff
@@ -59,7 +59,14 @@ public static void setupTimeoutServer() throws IOException {
 			}
 		});
 		server.rxListen().blockingGet();
+	}
 
+	@Test
+	public void testHugeBatch() throws IOException {
```
Comment (npomaroli) [inline tests/tests-core/src/main/java/com/gentics/mesh/search/ElasticSearchProviderTimeoutTest.java:65]: What does this test actually prove? There is no assertion in here - apart from that creating 1200 users succeeds, but that is completely unrelated to ES indexing.

File: elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java [hunk 4]
```diff
@@ -85,12 +89,22 @@ public void drain() {
...
 						timer.stop();
 						log.trace("Emitting bulk of size {} to subscriber", bulkableRequests.size());
-						BulkRequest request = new BulkRequest(bulkableRequests.asList());
-						bulkableRequests.clear();
-						if (log.isDebugEnabled()) {
-							log.debug("Sending bulk to elasticsearch:\n{}", request);
-						}
-						subscriber.onNext(request);
+						do {
+							Long[] bulkLength = new Long[] { 0L };
```
Comment (npomaroli) [inline elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java:93]: Use an AtomicLong here

File: elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java [hunk 5]
```diff
@@ -85,12 +89,22 @@ public void drain() {
...
+								.map(Optional::get)
+								.peek(bulkable -> bulkLength[0] += bulkable.bulkLength())
+								.collect(Collectors.toList());
+							BulkRequest request = new BulkRequest(requests);
+							if (log.isDebugEnabled()) {
+								log.debug("Sending bulk to elasticsearch:\n{}", request);
+							}
+							subscriber.onNext(request);
+						} while (!bulkableRequests.isEmpty());
 						BackpressureHelper.produced(requested, 1);
```
Comment (npomaroli) [inline elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java:108]: Doesn't this have to be changed as well, or at least moved to the do-while loop?

**Review by npomaroli — CHANGES_REQUESTED** [review]
We definitely should have a test case, which shows the error (without the fix) and proves that the fix works.

File: tests/tests-core/src/main/java/com/gentics/mesh/search/ElasticSearchProviderTimeoutTest.java (hunk 3 above)
Comment (plyhun) [reply tests/tests-core/src/main/java/com/gentics/mesh/search/ElasticSearchProviderTimeoutTest.java:65]: It is created during a timeout, so the indexing process should remain stable. Again, the direct use case is not reproducible, due to its spontaneous nature.

Comment (plyhun) [reply tests/tests-core/src/main/java/com/gentics/mesh/search/ElasticSearchProviderTimeoutTest.java:65]: I'll check if I'm able to extend BulkOperatorTest. Which will still be kinda artificial...

Comment (plyhun) [reply tests/tests-core/src/main/java/com/gentics/mesh/search/ElasticSearchProviderTimeoutTest.java:65]: Have made it eventually, with an extension that I don't really like.
